File: algorithms/search/binary/python/binary_search.py
"""
>>> arr = [1, 2, 3, 4, 5]
>>> binary_search(arr, 2)
True
>>> arr = [1, 2, 3, 4]
>>> binary_search(arr, 3)
True
>>> arr = [1, 2, 3, 4]
>>> binary_search(arr, 1)
True
>>> arr = [1, 2, 3, 4]
>>> binary_search(arr, 4)
True
>>> arr = [2, 4, 6, 8, 10]
>>> binary_search(arr, 3)
False
>>> arr = [2, 4, 6, 8, 10, 12]
>>> binary_search(arr, 3)
False
"""


def binary_search(arr, x):
    if len(arr) == 0:
        return False
    base = len(arr) / 2
    if arr[base] == x:
        return True
    if arr[base] > x:
        return binary_search(arr[:base], x)
    return binary_search(arr[base+1:], x)


if __name__ == "__main__":
    import doctest
    doctest.testmod()

File: algorithms/search/binary/C/binary_search.c
#include <stdio.h>
#include <stdlib.h>

int binarySearch(int *arr, int begin, int end, int x) {
    if (begin > end)
        return 0;
    int base = begin + (end - begin) / 2;
    if (arr[base] == x)
        return 1;
    if (arr[base] > x)
        return binarySearch(arr, begin, base-1, x);
    return binarySearch(arr, base+1, end, x);
}

int main() {
    int x; int n; int *arr;
    scanf("%d", &x);
    scanf("%d", &n);

    arr = (int*) malloc((n + 1) * sizeof(int));
    for (int i = 0; i < n; i++) {
        int temp = 0;
        scanf("%d", &temp);
        arr[i] = temp;
    }

    int found = binarySearch(arr, 0, n-1, x);
    if (found) {
        printf("%d found in array", x);
    } else {
        printf("%d not found in array", x);
    }

    return 0;
}
